### ISSUE TITLE
`as.matrix.projection()` for GAMs: Duplicated linear term

### ISSUE THREAD

**Issue by fweber144** [stan-dev/projpred#150]
On branch `develop` (commit 96d8e11d6f0ad0a1f7836c51700cb676b364bee5), this reprex:
```r
options(mc.cores = parallel::detectCores(logical = FALSE))
data("kidiq", package = "rstanarm")
kidiq <- head(kidiq, 45)
fit_gauss <- rstanarm::stan_gamm4(
  kid_score ~ s(mom_iq),
  data = kidiq,
  iter = 500,
  seed = 734572
)

library(projpred)
prj <- project(fit_gauss, solution_terms = c("mom_iq", "s(mom_iq)"),
               ndraws = 25)
colnames(as.matrix(prj))
```
gives
```
 [1] "b_Intercept"   "b_mom_iq"      "b_s(mom_iq).1" "b_s(mom_iq).2"
 [5] "b_s(mom_iq).3" "b_s(mom_iq).4" "b_s(mom_iq).5" "b_s(mom_iq).6"
 [9] "b_s(mom_iq).7" "b_s(mom_iq).8" "b_s(mom_iq).9" "sigma"
```
(so the output matrix includes a linear term for the smoothed predictor) whereas
```
colnames(as.matrix(fit_gauss))
```
(correctly) gives
```
 [1] "(Intercept)"           "s(mom_iq).1"          
 [3] "s(mom_iq).2"           "s(mom_iq).3"          
 [5] "s(mom_iq).4"           "s(mom_iq).5"          
 [7] "s(mom_iq).6"           "s(mom_iq).7"          
 [9] "s(mom_iq).8"           "s(mom_iq).9"          
[11] "sigma"                 "smooth_sd[s(mom_iq)1]"
[13] "smooth_sd[s(mom_iq)2]"
```
Perhaps the problem is not really with `as.matrix.projection()`, but rather with `project()` which would need to project onto the submodel with the smoothed term only (and not an additional linear term), even if `solution_terms = c("mom_iq", "s(mom_iq)")`?

There's another issue with `as.matrix.projection()` for GAMs visible here (the `smooth_sd` parameters) which I'll open a new issue for.

**Comment by fweber144** [comment]
> There's another issue with as.matrix.projection() for GAMs visible here (the smooth_sd parameters) which I'll open a new issue for.

This is issue #151 now.

**Comment by AlejandroCatalina** [comment]
This has to do with the way we split the original fit object's formula to do the forward search. We defined some conventions that seemed reasonable but they may give raise to stuff like this. In this case, for instance, it was decided that we also want to project the main effect of the smoothing term, which serves as a way of asking "do I need the smoothing term or is the linear term enough in terms of predictive accuracy?".

**Comment by AlejandroCatalina** [comment]
As far as I remember if the smooth is included first the linear term can’t be added, but if the linear term is added first the smooth can be included later. So a model with both could exist and from the modeling perspective it could make sense in the right model, but I agree they have to be properly constrained and so on. Probably @

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by fweber144** [comment]
The models I have tested with rstanarm and mgcv had the linear term included first, so that order didn't help.

**Comment by fweber144** [comment]
I have edited the [comment](https://github.com/stan-dev/projpred/issues/150#issuecomment-874203871) above.

**Comment by AlejandroCatalina** [comment]
Interestingly, if the linear term appears earlier and the smooth appears in solution terms that means that the projection actually performs better with both. We may need to add diagnostics that catch these potential issues in the projections.

**Comment by fweber144** [comment]
Sorry, I don't understand what you mean by

> the linear term appears earlier and the smooth appears in solution terms

Could you explain that?

**Comment by AlejandroCatalina** [comment]
Yes.

The way we coded the forward search implies that the only case where both the smooth and the linear term would appear in the solution is if the linear term is chosen first, indicating that it is a good predictor even without the smooth. Now it is possible that the smooth adds extra predictive power and therefore be chosen as part of the solution as well. I will check if this is indeed the case, that is, if including the smooth first actually removes the linear term. If this is the case, then what may happen is that the smooth may or may not be part of the optimal subset of predictors, and in this case it is fully possible for the projection including both terms to be unstable and have convergence issues, which is something we can add for improvement.

**Comment by fweber144** [comment]
Yes, that's the search behavior I would have expected, although I think the linear term needs to be removed *whenever* the smoothed term is included, not only

> if including the smooth first actually removes the linear term

Anyway, as I said above, I'm only referring to `project()` here in this issue, not to the search. I don't know if there are similar issues with the search, but I think it's already within `project()` that projpred should catch the case where the user supplies both (the linear as well as the smoothed term) to `solution_terms`.

**Comment by avehtari** [comment]
> In any case, a submodel shouldn't contain both, the nonsmoothed term (e.g. mom_iq) as well as the smoothed term (e.g. s(mom_iq)), right? 

It depends on whether s() term includes a linear term inside. There may have been confusion as I had thought s() would not contain the linear term, but based on this comment it does. I've also checked this at some point, but can't remember the details and the mgcv documentation doesn't say it clearly.

**Comment by fweber144** [comment]
Yes, that's true. The mgcv documentation is a bit scarce with regard to the mathematical details of the smoothing terms (e.g., `s()`). However, `?mgcv::gam.models` and the beginning of chapter 18 of ["The R book"](https://dx.doi.org/10.1002/9781118448908) by Michael Crawley make me think that an additional linear term is not necessary when using `s()` (so the typical usage of `s()` would be `s(x)` and not `s(x) + x`).

**Comment by fweber144** [comment]
The example at the beginning of section 5.2 of Wood (2006), together with the notation introduced earlier in the book, makes it clear that an additional linear term is not needed.

**References:**

Wood, S. N. (2006). Generalized additive models: An introduction with R (1st ed.). Chapman & Hall/CRC.
